### PR TITLE
Seq type

### PR DESCRIPTION
**Pull request by chentong319**
Add the support of sequence type in onnx. Sequence is a list of tensors which may have different shape. The current implementation keeps the type info of all members, meanwhile a member function for the first type is provided.  While the work on the lowering going on, this issue may be resolved.

Here is the test function added to check-onyx-lit:
```
  func @check_seq(%arg0: tensor<10x20xf32>, %arg1: tensor<5x20xf32>) -> tensor<*xf32> {
    %0 = "onnx.Constant"() {value = dense<0> : tensor<1xi32>} : () -> tensor<1xi32>
    %1 = "onnx.SequenceConstruct"(%arg0, %arg1) : (tensor<10x20xf32>, tensor<5x20xf32>) -> !onnx.Seq<tensor<10x20xf32>, tensor<5x20xf32>>
    %2 = "onnx.SequenceAt"(%1, %0) : (!onnx.Seq<tensor<10x20xf32>, tensor<5x20xf32>>, tensor<1xi32>) -> tensor<*xf32>
    return %2 : tensor<*xf32>
  }
```

### CODE REVIEW

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 1]
```diff
@@ -484,18 +484,48 @@ ONNXOpsDialect::ONNXOpsDialect(mlir::MLIRContext *ctx)
...
+  if (keyword == "Seq") {
+    if (parser.parseLess())
+      return Type();
+
+    SmallVector<mlir::Type, 1> elementTypes;
+    do {
+      llvm::SMLoc typeLoc = parser.getCurrentLocation();
+      mlir::Type elementType;
+      if (parser.parseType(elementType))
+        return nullptr;
```
Comment (tjingrant) [inline src/Dialect/ONNX/ONNXOps.cpp:506]: I tink you mean `return Type();`?

**Comment by tjingrant** [comment]
I have one high-level question: it seems like sequence right now is restricted to types whose length & component shapes are known at compile-time, is this going to be an issue down the road? I.e., any need/plans to support manipulating sequence within control flow operations?

File: src/Dialect/ONNX/ONNXOps.cpp (hunk 1 above)
Comment (chentong319) [reply src/Dialect/ONNX/ONNXOps.cpp]: yes. fixed.

**Comment by chentong319** [comment]
The sequence type is simply to record the type of tensor in the sequence. In current MLIR, the element type should be known, but not necessarily the shape. There are discussion in MLIR about supporting tensor with unknown element type. I feel that we can and have to rely on that extension.

**Pull request merged**